<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span><ul class="toc-item"><li><span><a href="#Подключение-библиотек,-чтение-и-изучение-файла" data-toc-modified-id="Подключение-библиотек,-чтение-и-изучение-файла-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Подключение библиотек, чтение и изучение файла</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Поиск-и-определение-пропусков-в-данных" data-toc-modified-id="Поиск-и-определение-пропусков-в-данных-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Поиск и определение пропусков в данных</a></span></li><li><span><a href="#Обработка-артефактов-в-датасете" data-toc-modified-id="Обработка-артефактов-в-датасете-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обработка артефактов в датасете</a></span><ul class="toc-item"><li><span><a href="#Устранение-дубликатов-записанных-в-разных-регистрах" data-toc-modified-id="Устранение-дубликатов-записанных-в-разных-регистрах-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Устранение дубликатов записанных в разных регистрах</a></span></li><li><span><a href="#Исправление-некорректных-значений-в-столбце-трудового-стажа-'days_employed'" data-toc-modified-id="Исправление-некорректных-значений-в-столбце-трудового-стажа-'days_employed'-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>Исправление некорректных значений в столбце трудового стажа 'days_employed'</a></span></li><li><span><a href="#Обзор-и-поиск-возможных-ошибок-в-столбце-'children'" data-toc-modified-id="Обзор-и-поиск-возможных-ошибок-в-столбце-'children'-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>Обзор и поиск возможных ошибок в столбце 'children'</a></span></li><li><span><a href="#Отсуствующий-возраст-клиента-в-столбце-'dob_years" data-toc-modified-id="Отсуствующий-возраст-клиента-в-столбце-'dob_years-2.2.4"><span class="toc-item-num">2.2.4&nbsp;&nbsp;</span>Отсуствующий возраст клиента в столбце 'dob_years</a></span></li><li><span><a href="#Проверка-на-артефакты-других-столбцов-данных" data-toc-modified-id="Проверка-на-артефакты-других-столбцов-данных-2.2.5"><span class="toc-item-num">2.2.5&nbsp;&nbsp;</span>Проверка на артефакты других столбцов данных</a></span></li></ul></li><li><span><a href="#Заполнение-пропусков-медианными-значениями-по-категории" data-toc-modified-id="Заполнение-пропусков-медианными-значениями-по-категории-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Заполнение пропусков медианными значениями по категории</a></span></li><li><span><a href="#Замена-типа-данных" data-toc-modified-id="Замена-типа-данных-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Замена типа данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Лемматизация</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Результаты-исследования" data-toc-modified-id="Результаты-исследования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Результаты исследования</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Есть ли зависимость между наличием детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li></ul></li><li><span><a href="#Общие-выводы" data-toc-modified-id="Общие-выводы-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общие выводы</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Обзор данных 

### Подключение библиотек, чтение и изучение файла

In [998]:
import pandas as pd
import numpy as np
from pymystem3 import Mystem
from  collections import Counter

In [999]:
data_bank_clients = pd.read_csv('/datasets/data.csv')

Ознакомимся с данными, выведем первые 10 строк датасета

In [1000]:
data_bank_clients.head(10) 

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Оценим корректность наименований столбцов

In [1001]:
data_bank_clients.columns.tolist()

['children',
 'days_employed',
 'dob_years',
 'education',
 'education_id',
 'family_status',
 'family_status_id',
 'gender',
 'income_type',
 'debt',
 'total_income',
 'purpose']

Выведем на экран общую информацию о датасете

In [1002]:
data_bank_clients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Выведем на экран сводную статистику по датафрему

In [1003]:
data_bank_clients.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

Согласно документации в датафрейме содержится следующая информация:

* children — количество детей в семье
* days_employed — общий трудовой стаж в днях
* dob_years — возраст клиента в годах
* education — уровень образования клиента
* education_id — идентификатор уровня образования
* family_status — семейное положение
* family_status_id — идентификатор семейного положения
* gender — пол клиента
* income_type — тип занятости
* debt — имел ли задолженность по возврату кредитов
* total_income — ежемесячный доход
* purpose — цель получения кредита

**ВЫВОДЫ ПО ГЛАВЕ 1. Обзор данных**

1. Исходный файл содержит 21525 строк и 12 столбцов.
2. Тип данных по столбцам соответствует информации в них. Для удобства работы тип данных в столбцах days_employed и total_income нужно привести к целочисленному типу.
3. Наименования колонок соответствуют "хорошему" стилю и их названия соответствуют документации. 

**Проблемные вопросы по исходному датасету:**

* В колонке 'days_employed' имеются отрицательные значения и неудобные для анализа значения. Имеются значения не соответствующие возрасту клиента.
* Пропущенные значения имеются в колонках 'days_employed' и 'total_income' (для каждой колонки заполнено 19351 строк из 21525).
* В колонке 'education' данные представлены в разном регистре.
* В колонке 'children' присутствуют отрицательные значения, а максимальное количество детей у клиента - 20.
* В колонке dob_years (возраст клиента) встречается возраст '0'. Возраст клиента не соответствует значению стажа.

## Предобработка данных

### Поиск и определение пропусков в данных

Проверим общее количество пропусков в каждой колонке.

In [1004]:
isna_dbc = data_bank_clients.isna().sum()
isna_dbc

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Рассчитаем процентное отношение пропущенных значений к общему числу данных по столбцам 'days_employed' и 'total_income'

In [1005]:
isna_de = isna_dbc['days_employed'] / len(data_bank_clients)
isna_ti = isna_dbc['total_income'] / len(data_bank_clients)
print("Процент пропущенных значений по колонке 'days_employed': {:.1%}".format(isna_de))
print("Процент пропущенных значений по колонке 'total_income': {:.1%}".format(isna_ti))

Процент пропущенных значений по колонке 'days_employed': 10.1%
Процент пропущенных значений по колонке 'total_income': 10.1%


Убедимся что пропуски в обоих столбцах находятся на одних и тех же строках: получим таблицу только их строк с пропущенными значениями, выведем краткую информацию

In [1006]:

data_bank_clients.query('days_employed.isna()').info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2174 entries, 12 to 21510
Data columns (total 12 columns):
children            2174 non-null int64
days_employed       0 non-null float64
dob_years           2174 non-null int64
education           2174 non-null object
education_id        2174 non-null int64
family_status       2174 non-null object
family_status_id    2174 non-null int64
gender              2174 non-null object
income_type         2174 non-null object
debt                2174 non-null int64
total_income        0 non-null float64
purpose             2174 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 220.8+ KB


Там где пустые значения 'days_employed', там  и  пустые 'total_income' и всего 2174.

***Трудовой стаж и ежемесячный доход - важные критерии для выдачи кредита, особенно при большой сумме заема, например при приобретении недвижижмости. Маленькая вероятность того, что изначально эти данные не были заполнены.***

Возможно, при выгрузке данных произошел сбой и пропуски появились по определенному приниципу (например только у клиентов со средним образованием или у клиентов определенного семейного положения) или по причине технического сбоя при занесении данных.

Проверим зависимость появления пропусков от других данных представленных в исходной таблице.  
null_persent_category(column) - группирует по категорям столбца column и возвращает процент пропусков по категории.  
Столбец purpose выбран для подсчета значение т.к. в нем нет пропусков и он не будет использоваться для группировки по категориям в текущем разделе.

In [1007]:
def null_persent_category(column):
    full_category = data_bank_clients.pivot_table(index=[column], 
                                                  values=['purpose', 'days_employed'], aggfunc=['count'])
    full_category['percent_null'] = (1 - (full_category['count']['days_employed'] / 
                                          full_category['count']['purpose'])) * 100
    full_category['percent_null'] = round(full_category['percent_null'], 1)
    return full_category

Проверим работу функции на примере со столбцом 'children', на случай ошибки печатаем 'WARNING! Неподходящие столбцы'

In [1008]:
try:
    display(null_persent_category('children'))
except:
    print('WARNING! Неподходящие столбцы')

count         percent_null
         days_employed purpose             
children                                   
-1                  44      47          6.4
 0               12710   14149         10.2
 1                4343    4818          9.9
 2                1851    2055          9.9
 3                 294     330         10.9
 4                  34      41         17.1
 5                   8       9         11.1
 20                 67      76         11.8

- Пропуски разбросаны равномерно по всем категориям и процент пропусков по категориям находится в основном в диапазоне от 9,9 до 11,8.   
- Ислючения составляют данные о клиентах с 4 детьми - здесь процент пропусков в данных выше -17,1%, при этом количество таких клиентов в выборке мало - 41 из 21525.
- Очень мала выборка по клиетам с 5 детьми - 9 из 21525.
- Процент пропусков у клиентов со значением -1 не оцениваю, так как значение ошибочно.

In [1009]:
try:
    display(null_persent_category('education'))
except:
    print('WARNING! Неподходящие столбцы')

count         percent_null
                    days_employed purpose             
education                                             
ВЫСШЕЕ                        251     274          8.4
Высшее                        243     268          9.3
НАЧАЛЬНОЕ                      16      17          5.9
НЕОКОНЧЕННОЕ ВЫСШЕЕ            22      29         24.1
Начальное                      14      15          6.7
Неоконченное высшее            40      47         14.9
СРЕДНЕЕ                       705     772          8.7
Среднее                       646     711          9.1
УЧЕНАЯ СТЕПЕНЬ                  1       1          0.0
Ученая степень                  1       1          0.0
высшее                       4222    4718         10.5
начальное                     231     250          7.6
неоконченное высшее           613     668          8.2
среднее                     12342   13750         10.2
ученая степень                  4       4          0.0

- Пропуски встречаются во всех категориях стролбца образование, за исключением клиентов с ученой степенью.  
- На первый взгляд есть явно высокий процент пропусков - 24 % у клиентов с неоконченным высшим образованием и явно низкий - у клиентов с начальным образованием. Но сейчас сложно оценить, так как в столбце образование данные не приведены к единому регистру и группировка не дала окончательный ответ.  

Оценим пропуски в категорях по столбцу 'family_status'

In [1010]:
try:
    display(null_persent_category('family_status'))
except:
    print('WARNING! Неподходящие столбцы')

count         percent_null
                      days_employed purpose             
family_status                                           
Не женат / не замужем          2525    2813         10.2
в разводе                      1083    1195          9.4
вдовец / вдова                  865     960          9.9
гражданский брак               3735    4177         10.6
женат / замужем               11143   12380         10.0

- Пропуски по категориям столбца семейного положения распределены равномерно ~10% и присутствуют во всех категориях.  

оценим пропуски в категорях по столбцу 'gender'

In [1011]:
try:
    display(null_persent_category('gender'))
except:
    print('WARNING! Неподходящие столбцы')

count         percent_null
       days_employed purpose             
gender                                   
F              12752   14236         10.4
M               6598    7288          9.5
XNA                1       1          0.0

- Еще одна проблема в данных - третий пол 'XNA'. К счастью, такая строчка одна и это не повлияет на результат исследования. До уточнения этого вопроса оставим значение таким, какое оно есть.
- Пропуски распределены в том же диапазоне, что и в общей выборке ~10%.  

оценим пропуски в категорях по столбцу'income_type'

In [1012]:
try:
    display(null_persent_category('income_type'))
except:
    print('WARNING! Неподходящие столбцы')

count         percent_null
                days_employed purpose             
income_type                                       
безработный                 2       2          0.0
в декрете                   1       1          0.0
госслужащий              1312    1459         10.1
компаньон                4577    5085         10.0
пенсионер                3443    3856         10.7
предприниматель             1       2         50.0
сотрудник               10014   11119          9.9
студент                     1       1          0.0

- Явная проблема с предпринимателями: из-за того, что выборка по ним слишком мала (2 клиента) процент пропусков в данных составил 50%.
- По категориям типа дохода, представленным в датасете в достаточном количестве процент пропусков находится на уровне среднего по всему столбцу ~10%.

In [1013]:
try:
    display(null_persent_category('debt'))
except:
    print('WARNING! Неподходящие столбцы')

count         percent_null
     days_employed purpose             
debt                                   
0            17780   19784         10.1
1             1571    1741          9.8

- Группировка в соответствии с кредитной историей клиентов показывает равномерное распределение пропусков на уровне ~10%.

**Вывод**
***

Итак, пропущенные значения имеются в двух колонках: 'days_employed' - трудовой стаж и 'total_income' - ежемесячный доход. Пропуски в колонках находятся в одних и тех же строках - 2174 строк.  

Рассмотрение распределения пропусков в соответствии с различными группами показало, что пропуски распределены внутри категорий столбцов на том же уровне (~10%), что и в общей выборке, за исключением случаев, когда выборка по категории слишком мала. Какая-либо закономерность появления пропущенных значений не прослеживается.

Для последующей обработки пропусков необходимо обработать артефакты в данных.

Так же, был обнаружен еще один артефакт: значение XNA - в столбце пола.

### Обработка артефактов в датасете

В исходном файле имеются некорректные данные, которые усложняют дальнейшую обработку:
- категориальные данные записаны в разных регистрах
- отрицательные значения стажа клиентов
- максимальное значение стажа клиента в пересчете на года - 1100 лет
- отрицательное количество детей
- максимальное количество детей 20
- возраст клиентов 0 лет
- пол 'XNA'

 #### Устранение дубликатов записанных в разных регистрах

Составим список из столбцов, в которых могла появится проблема с форматом данных

In [1014]:
name_columns_df_bc = ['education', 'family_status', 'gender', 'income_type', 'debt', 'purpose']

for column in name_columns_df_bc:
    print(f'Уникальные значения столбца {column}:')
    print(*(data_bank_clients[column].unique()), sep='\n')
    print()

Уникальные значения столбца education:
высшее
среднее
Среднее
СРЕДНЕЕ
ВЫСШЕЕ
неоконченное высшее
начальное
Высшее
НЕОКОНЧЕННОЕ ВЫСШЕЕ
Неоконченное высшее
НАЧАЛЬНОЕ
Начальное
Ученая степень
УЧЕНАЯ СТЕПЕНЬ
ученая степень

Уникальные значения столбца family_status:
женат / замужем
гражданский брак
вдовец / вдова
в разводе
Не женат / не замужем

Уникальные значения столбца gender:
F
M
XNA

Уникальные значения столбца income_type:
сотрудник
пенсионер
компаньон
госслужащий
безработный
предприниматель
студент
в декрете

Уникальные значения столбца debt:
0
1

Уникальные значения столбца purpose:
покупка жилья
приобретение автомобиля
дополнительное образование
сыграть свадьбу
операции с жильем
образование
на проведение свадьбы
покупка жилья для семьи
покупка недвижимости
покупка коммерческой недвижимости
покупка жилой недвижимости
строительство собственной недвижимости
недвижимость
строительство недвижимости
на покупку подержанного автомобиля
на покупку своего автомобиля
операции с коммерческой

Приведем данные в колонках 'education', 'family_status', 'income_type' 'purpose' к одному регистру

In [1015]:
name_columns_dbc_object = ['education', 'family_status', 'income_type', 'purpose']
for column in name_columns_dbc_object:
    data_bank_clients[column] = data_bank_clients[column].str.lower()

Проверим результат смены регистра: выведем уникальные значения на экран

In [1016]:
for column in name_columns_dbc_object:
    print(f'Уникальные значения столбца {column}:')
    print(data_bank_clients[column].unique())
    print()

Уникальные значения столбца education:
['высшее' 'среднее' 'неоконченное высшее' 'начальное' 'ученая степень']

Уникальные значения столбца family_status:
['женат / замужем' 'гражданский брак' 'вдовец / вдова' 'в разводе'
 'не женат / не замужем']

Уникальные значения столбца income_type:
['сотрудник' 'пенсионер' 'компаньон' 'госслужащий' 'безработный'
 'предприниматель' 'студент' 'в декрете']

Уникальные значения столбца purpose:
['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'операции со своей недвижимостью' 'автомобили' 'заняться об

**Вывод**

Категориальные значения в столбцах: 'education', 'family_status', 'income_type', 'purpose' приведены к единому регистру.

 #### Исправление некорректных значений в столбце трудового стажа 'days_employed'

Сгруппируем данные по столбцу тип занятости 'income_type' и выведем среднее значение стажа заемщика.

In [1017]:
days_employed_mean = data_bank_clients.groupby(by=['income_type'])['days_employed'].mean()
days_employed_mean

income_type
безработный        366413.652744
в декрете           -3296.759962
госслужащий         -3399.896902
компаньон           -2111.524398
пенсионер          365003.491245
предприниматель      -520.848083
сотрудник           -2326.499216
студент              -578.751554
Name: days_employed, dtype: float64

In [1018]:
data_bank_clients['days_employed'].describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

Две проблемы со столбцом 'days_employed':
- Трудовой стаж безработных и пенсионеров записан не в днях или с неким коэффициентом.
- Трудовой стаж работающих людей отрицательный.

***Разберемся сначала с порядком цифр для безработных и пенсионеров.***

In [1019]:
data_pensionika = data_bank_clients[data_bank_clients['income_type'] == 'пенсионер']

# выводим уникальные значения возраста для пенсионеров
print(sorted(data_pensionika['dob_years'].unique()))

# таблица для пенсионеров возраст которых меньше 40 лет
data_pensionika[data_pensionika['dob_years'] < 40]

[0, 22, 24, 26, 27, 28, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74]


children  days_employed  dob_years education  education_id  \
99            0  346541.618895          0   среднее             1   
157           0  348414.028009         38   среднее             1   
578           0  397856.565013          0   среднее             1   
776           0  365336.560325         38   среднее             1   
1175          0  370879.508002          0   среднее             1   
...         ...            ...        ...       ...           ...   
19439         0  389397.167577         26    высшее             0   
20271         0  383832.244324         37    высшее             0   
20462         0  338734.868540          0   среднее             1   
20560         0  333394.000848         33   среднее             1   
20577         0  331741.271455          0   среднее             1   

               family_status  family_status_id gender income_type  debt  \
99           женат / замужем                 0      F   пенсионер     0   
157          женат / замужем                 0      F   пенсионер     1   
578          женат / замужем                 0      F   пенсионер     0   
776          женат / замужем                 0      F   пенсионер     0   
1175         женат / замужем                 0      F   пенсионер     0   
...                      ...               ...    ...         ...   ...   
19439        женат / замужем                 0      F   пенсионер     0   
20271        женат / замужем                 0      F   пенсионер     0   
20462        женат / замужем                 0      F   пенсионер     0   
20560        женат / замужем                 0      M   пенсионер     0   
20577  не женат / не замужем                 4      F   пенсионер     0   

        total_income                                 purpose  
99      71291.522491                              автомобиль  
157    113560.650035                    сделка с автомобилем  
578     97620.687042  строительство собственной недвижимости  
776     73859.425084                    покупка недвижимости  
1175   313949.845188   получение дополнительного образования  
...              ...                                     ...  
19439  214963.301941                    покупка недвижимости  
20271  188236.765814                   на покупку автомобиля  
20462  259193.920299                    покупка своего жилья  
20560  163411.879628                             образование  
20577  129788.762899                            недвижимость  

[62 rows x 12 columns]

In [1020]:
# рассчитаем максимальный трудовой стаж в днях для пенсионеров
# для расчета обозначим, что трудовая деятельность начинается в 18 лет, выход на пенсию - 65 лет.

long_emploed = (65 - 18) * 365
mid_age_pens = data_bank_clients[data_bank_clients['income_type'] == 'пенсионер']['dob_years'].mean()
print(f'Если работать {65-18} лет без перерывов, то это {long_emploed} дней')
print(f'А средний стаж пенсионера по исходным данным {int(days_employed_mean[4] / 365)} лет и это {int(days_employed_mean[4])} дней')
print(f'При этом средний возраст пенсионеров {int(mid_age_pens)} лет')

Если работать 47 лет без перерывов, то это 17155 дней
А средний стаж пенсионера по исходным данным 1000 лет и это 365003 дней
При этом средний возраст пенсионеров 59 лет


In [1021]:
# выведем на экран максимальное и минимальное значение стажа для пенсионеров в днях
display(data_bank_clients[data_bank_clients['income_type'] == 'пенсионер']['days_employed'].min())
data_bank_clients[data_bank_clients['income_type'] == 'пенсионер']['days_employed'].max()

328728.72060451825

401755.40047533

Можно предположить, что данные по стажу для пенсионеров записаны не в днях, а в часах. Проверим эту гипотезу и разделим средний стаж пенсионера на 24 часа.

In [1022]:
# выведем на экран средний трудовой стаж исходя из гипотезы что он записан в часах
print('Средний стаж пенсионера теперь - ',int(days_employed_mean[4] / 365 / 24), 'год')

Средний стаж пенсионера теперь -  41 год


Это уже ближе к истине! Переведем значения из формата часов в дни в исходном датафрейме.

In [1023]:
# разделить значения в столбце 'days_employed' на 24 
data_bank_clients.loc[data_bank_clients['income_type'] == 'пенсионер', 'days_employed'] /= 24

# проверим результат - выведем среднее значение трудового стажа по професссиям
data_bank_clients.groupby(by=['income_type'])['days_employed'].mean()

income_type
безработный        366413.652744
в декрете           -3296.759962
госслужащий         -3399.896902
компаньон           -2111.524398
пенсионер           15208.478802
предприниматель      -520.848083
сотрудник           -2326.499216
студент              -578.751554
Name: days_employed, dtype: float64

In [1024]:
# проверим, нет ли данных, в которых трудовой стаж будет превышать трудовой возраст (трудовой восраст от 18 лет)
data_bank_clients.query('income_type = пенсионер' and '(days_employed / 365) > dob_years')

children  days_employed  dob_years education  education_id  \
99            0   14439.234121          0   среднее             1   
157           0   14517.251167         38   среднее             1   
578           0   16577.356876          0   среднее             1   
751           0   16281.477669         41   среднее             1   
776           0   15222.356680         38   среднее             1   
...         ...            ...        ...       ...           ...   
20271         0   15993.010180         37    высшее             0   
20462         0   14113.952856          0   среднее             1   
20560         0   13891.416702         33   среднее             1   
20577         0   13822.552977          0   среднее             1   
20829         1   16591.942999         40   среднее             1   

               family_status  family_status_id gender income_type  debt  \
99           женат / замужем                 0      F   пенсионер     0   
157          женат / замужем                 0      F   пенсионер     1   
578          женат / замужем                 0      F   пенсионер     0   
751          женат / замужем                 0      M   пенсионер     0   
776          женат / замужем                 0      F   пенсионер     0   
...                      ...               ...    ...         ...   ...   
20271        женат / замужем                 0      F   пенсионер     0   
20462        женат / замужем                 0      F   пенсионер     0   
20560        женат / замужем                 0      M   пенсионер     0   
20577  не женат / не замужем                 4      F   пенсионер     0   
20829        женат / замужем                 0      F   пенсионер     0   

        total_income                                 purpose  
99      71291.522491                              автомобиль  
157    113560.650035                    сделка с автомобилем  
578     97620.687042  строительство собственной недвижимости  
751    151898.693438         операции со своей недвижимостью  
776     73859.425084                    покупка недвижимости  
...              ...                                     ...  
20271  188236.765814                   на покупку автомобиля  
20462  259193.920299                    покупка своего жилья  
20560  163411.879628                             образование  
20577  129788.762899                            недвижимость  
20829   39107.907534                   получение образования  

[71 rows x 12 columns]

Имеются данные, где трудовой стаж превышает возраст. Таких строк - 71.  
- В них присутствуют пенсионеры с отсутствующим возрастом (0) и пенсионеры от 22 лет. Касательно возраста, возможно, что человек вышел на пенсию раньше, например военный. Или пенсия была выдана по инвалидности. Вариантов много.  
- Стаж не должен превышать возраст человека. Возможно, в стаж попадают значения возраста клиента в днях или здесь возраст клиента указан неверно. До уточнения этих данных оставим значения как есть.

***Рассмотрим даннные о безработных***

In [1025]:
data_jobless = data_bank_clients[data_bank_clients['income_type'] == 'безработный']

# выводим уникальные значения возраста для безработных
print(sorted(data_jobless['dob_years'].unique()))

# проверим гипотезу, что значения записаны в часах.
data_jobless['days_employed'] / 365 / 24

[31, 45]


3133     38.530190
14798    45.125895
Name: days_employed, dtype: float64

В таком случае, стаж в одном случае больше возраста заемщика, в другом случае равен возрасту заемщика. Возможно, данные были записаны с ошибкой в плавающей точке.

In [1026]:
data_jobless['days_employed'] / 365 / 100

3133      9.247246
14798    10.830215
Name: days_employed, dtype: float64

При делении на коэффициент 100 трудовой стаж в 9 и 10 лет становится более вероятным. Возможно эти 2 человека не работают временно или работают не официально.

In [1027]:
# приведем трудовой стаж безработных к реальному значению
data_bank_clients.loc[data_bank_clients['income_type'] == 'безработный', 'days_employed'] /= 100
data_bank_clients[data_bank_clients['income_type'] == 'безработный']

children  days_employed  dob_years education  education_id  \
3133          1    3375.244668         31   среднее             1   
14798         0    3953.028387         45    высшее             0   

          family_status  family_status_id gender  income_type  debt  \
3133    женат / замужем                 0      M  безработный     1   
14798  гражданский брак                 1      F  безработный     0   

        total_income                  purpose  
3133    59956.991984  покупка жилья для сдачи  
14798  202722.511368             ремонт жилью

**Отрицательные значения в столбце 'days_employed'**

In [1028]:
# выведем на экран среднее значения по отрицательному стажу работников
# разделим среднее на 365 чтобы результат получился в годах для наглядности
data_bank_clients[data_bank_clients['days_employed'] < 0]['days_employed'].mean() / 365

-6.446618991777744

Средний стаж всех категорий работников с отрицательным значением равен 6 годам. Из этого, можно сделать вывод, что стаж работающих клиентов записывается в отрицательном значении.  
Судя по тому, что стаж заещика записан в днях, возможно, подсчет ведется автоматически из даты начала трудового стажа и текущей даты(или окончания) и даты либо были сразу указаны в обратном порядке или вычитается дата окончания из даты начала.

In [1029]:
# заменим отрицательные значения на положительные
data_bank_clients.loc[data_bank_clients['days_employed'] < 0, 'days_employed'] = abs(data_bank_clients['days_employed'])

In [1030]:
# проверим как изменился средний стаж по группам работников
data_bank_clients.groupby(by=['income_type'])['days_employed'].mean()

income_type
безработный         3664.136527
в декрете           3296.759962
госслужащий         3399.896902
компаньон           2111.524398
пенсионер          15208.478802
предприниматель      520.848083
сотрудник           2326.499216
студент              578.751554
Name: days_employed, dtype: float64

За счет того, что в выборке имеется только 2 безработных клиента, их средний общий трудовой стаж выглядит не маленьким.

**Вывод**

До уточнения данных по столбцу 'days_employed' отрицательные значения переведены в положительные, а стаж в 1000+ лет переведен с коэффициентом на более реальный.

 #### Обзор и поиск возможных ошибок в столбце 'children'

In [1031]:
data_bank_clients['children'].describe()

count    21525.000000
mean         0.538908
std          1.381587
min         -1.000000
25%          0.000000
50%          0.000000
75%          1.000000
max         20.000000
Name: children, dtype: float64

Рассмотрим два сомнительных значения в столбце 'children':
- Отрицательное количество детей.
- Максимальное количество детей - 20.

In [1032]:
# посмотрим сколько строк с отрицательным значением и значением 20
data_bank_clients['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Имеется 47 клиентов банка с -1 ребенком и 76 клиентов с 20 детьми. Что выглядит в первом случае нереальным, а во втором - крайне маловероятным.


In [1033]:
# ознакомимся с данными о клиентах с 20 детьми
data_bank_clients[data_bank_clients['children'] == 20].head(10)

children  days_employed  dob_years education  education_id  \
606         20     880.221113         21   среднее             1   
720         20     855.595512         44   среднее             1   
1074        20    3310.411598         56   среднее             1   
2510        20    2714.161249         59    высшее             0   
2941        20    2161.591519          0   среднее             1   
3302        20            NaN         35   среднее             1   
3396        20            NaN         56    высшее             0   
3671        20     913.161503         23   среднее             1   
3697        20    2907.910616         40   среднее             1   
3735        20     805.044438         26    высшее             0   

              family_status  family_status_id gender  income_type  debt  \
606         женат / замужем                 0      M    компаньон     0   
720         женат / замужем                 0      F    компаньон     0   
1074        женат / замужем                 0      F    сотрудник     1   
2510         вдовец / вдова                 2      F    сотрудник     0   
2941        женат / замужем                 0      F    сотрудник     0   
3302  не женат / не замужем                 4      F  госслужащий     0   
3396        женат / замужем                 0      F    компаньон     0   
3671  не женат / не замужем                 4      F    сотрудник     0   
3697       гражданский брак                 1      M    сотрудник     0   
3735  не женат / не замужем                 4      M    сотрудник     0   

       total_income                                purpose  
606   145334.865002                          покупка жилья  
720   112998.738649                   покупка недвижимости  
1074  229518.537004                  получение образования  
2510  264474.835577  операции с коммерческой недвижимостью  
2941  199739.941398                  на покупку автомобиля  
3302            NaN                 профильное образование  
3396            NaN                     высшее образование  
3671  101255.492076     на покупку подержанного автомобиля  
3697  115380.694664     на покупку подержанного автомобиля  
3735  137200.646181                           ремонт жилью

In [1034]:
# ознакомимся с данными о клиентах с -1 детьми
data_bank_clients[data_bank_clients['children'] == -1].head(10)

children  days_employed  dob_years education  education_id  \
291         -1    4417.703588         46   среднее             1   
705         -1     902.084528         50   среднее             1   
742         -1    3174.456205         57   среднее             1   
800         -1   14582.827176         54   среднее             1   
941         -1            NaN         57   среднее             1   
1363        -1    1195.264956         55   среднее             1   
1929        -1    1461.303336         38   среднее             1   
2073        -1    2539.761232         42   среднее             1   
3814        -1    3045.290443         26   среднее             1   
4201        -1     901.101738         41   среднее             1   

              family_status  family_status_id gender  income_type  debt  \
291        гражданский брак                 1      F    сотрудник     0   
705         женат / замужем                 0      F  госслужащий     0   
742         женат / замужем                 0      F    сотрудник     0   
800   не женат / не замужем                 4      F    пенсионер     0   
941         женат / замужем                 0      F    пенсионер     0   
1363        женат / замужем                 0      F    компаньон     0   
1929  не женат / не замужем                 4      M    сотрудник     0   
2073              в разводе                 3      F    компаньон     0   
3814       гражданский брак                 1      F  госслужащий     0   
4201        женат / замужем                 0      F  госслужащий     0   

       total_income                          purpose  
291   102816.346412           профильное образование  
705   137882.899271          приобретение автомобиля  
742    64268.044444       дополнительное образование  
800    86293.724153       дополнительное образование  
941             NaN     на покупку своего автомобиля  
1363   69550.699692           профильное образование  
1929  109121.569013                    покупка жилья  
2073  162638.609373                    покупка жилья  
3814  131892.785435            на проведение свадьбы  
4201  226375.766751  операции со своей недвижимостью

Проверим нет ли взаимосвязи появления отрицательных значений с другими значениями, выведем уникальные значения возраста, и образования

In [1035]:
print(sorted(data_bank_clients[data_bank_clients['children'] == -1]['dob_years'].unique()))
print(sorted(data_bank_clients[data_bank_clients['children'] == -1]['education'].unique()))

[23, 26, 27, 28, 30, 31, 32, 33, 34, 35, 37, 38, 40, 41, 42, 43, 44, 46, 48, 50, 51, 53, 54, 55, 57, 59, 61, 63, 64, 69]
['высшее', 'неоконченное высшее', 'среднее']


Значение -1 попало к клиентам разного возраста от 23 до 69 лет и с разным образованием.  
Посчитаем данные ошибочными и примем, что 20 - это 2 ребенка, а -1 это 1.

In [1036]:
# заменим ошибочные значения
data_bank_clients['children'] = data_bank_clients['children'].replace(20, 2)
data_bank_clients['children'] = data_bank_clients['children'].replace(-1, 1)

In [1037]:
# проверим результат замены: выведем на экран уникальное количество детей
data_bank_clients['children'].unique()

array([1, 0, 3, 2, 4, 5])

**Вывод**

Отрицательное количество детей заменено на положительное, а значения 20 заменены на 2.

#### Отсуствующий возраст клиента в столбце 'dob_years

Рассмотрим подробнее все значения возраста клиентов.

In [1038]:
data_bank_clients['dob_years'].describe()

count    21525.000000
mean        43.293380
std         12.574584
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

Выведем в сводную таблицу значения возраста для каждой категории работников

In [1039]:
pivot_dbc = data_bank_clients.pivot_table(index=['income_type'], values='dob_years', aggfunc=['count', 'min', 'max', 'mean', 'median'])
pivot_dbc.reset_index()

income_type     count       min       max       mean    median
                   dob_years dob_years dob_years  dob_years dob_years
0      безработный         2        31        45  38.000000      38.0
1        в декрете         1        39        39  39.000000      39.0
2      госслужащий      1459         0        75  40.636737      40.0
3        компаньон      5085         0        74  39.697542      39.0
4        пенсионер      3856         0        74  59.063019      60.0
5  предприниматель         2        27        58  42.500000      42.5
6        сотрудник     11119         0        74  39.821027      39.0
7          студент         1        22        22  22.000000      22.0

Из таблицы видно, что минимальное значение возраста - 0 присутствует у госслужащих, пенсионеров, компаньонов и сотрудников.  
Возможно произошла ошибка при выгрузке данных.  
0 - значения возможно заполнить средним возратом по группам сотрудников.   

Оценим сколько строк с 0 возрастом присуствуют в таблице

In [1040]:
data_bank_clients[data_bank_clients['dob_years'] == 0]['dob_years'].count()

101

Всего 101 строка с 0 значениями возраста из 25525 строк (~ 0,4 %).  
То есть замена данных на среднее значение по категории повлияет на данные слабо, но для обработки будет проще без 0 значений.  

Сохраним средние значения по типу занятости для тех категорий, у которых минимальный возраст равен 0 c округлением до полных лет в меньшую сторону

In [1041]:
mean_age_state = int(pivot_dbc.loc['госслужащий']['mean'])
mean_age_partner = int(pivot_dbc.loc['компаньон']['mean'])
mean_age_pensioner = int(pivot_dbc.loc['пенсионер']['mean'])
mean_age_employee = int(pivot_dbc.loc['сотрудник']['mean'])

Сохраним в список категории клиентов для которых нужна замена 0-го возраста и среднее значение возраста для категории

In [1042]:
age_null_type = ['госслужащий', 'компаньон', 'пенсионер', 'сотрудник']
age_null_mean = [mean_age_state, mean_age_partner, mean_age_pensioner, mean_age_employee]                

Заменим циклом значения возраста с 0 на среднее по категории типа занятости
- замена только в строках с возрастом = 0, по значению категории

In [1043]:
for i in range(len(age_null_type)):
    data_bank_clients.loc[(data_bank_clients['dob_years'] == 0) & (data_bank_clients['income_type'] == age_null_type[i]),
                          'dob_years'] = age_null_mean[i]

Проверим остались ли категории работников со значением 0 в столбце возраст 'dob_years'

In [1044]:
data_bank_clients[data_bank_clients['dob_years'] == 0]['income_type'].unique()

array([], dtype=object)

In [1045]:
# проверка исправленных данных
data_bank_clients.pivot_table(index=['income_type'], values='dob_years', aggfunc=['min', 'mean']).reset_index()

income_type       min       mean
                   dob_years  dob_years
0      безработный        31  38.000000
1        в декрете        39  39.000000
2      госслужащий        19  40.801234
3        компаньон        19  39.850934
4        пенсионер        22  59.369035
5  предприниматель        27  42.500000
6        сотрудник        19  40.013940
7          студент        22  22.000000

**Вывод**

Минимального значения возраста '0' в таблице больше нет. Среднее значение возраста по категориям по типу занятости после замены '0' на среднее значение изменились слабо.

#### Проверка на артефакты других столбцов данных

In [1046]:
data_bank_clients.columns.tolist()

['children',
 'days_employed',
 'dob_years',
 'education',
 'education_id',
 'family_status',
 'family_status_id',
 'gender',
 'income_type',
 'debt',
 'total_income',
 'purpose']

Проверим на наличие артефактов столбцы, которым ранее внимание было не уделено.  
Это столбцы - 'education_id', 'family_status_id', 'gender', 'debt', 'purpose'.

In [1047]:
# выведем уникальные значения столбцов на экран 
columns = ['education', 'education_id', 'family_status', 'family_status_id', 'gender', 'debt', 'purpose']
for column in columns:
    print(f'Уникальные значения столбца {column}:')
    print(data_bank_clients[column].unique())

Уникальные значения столбца education:
['высшее' 'среднее' 'неоконченное высшее' 'начальное' 'ученая степень']
Уникальные значения столбца education_id:
[0 1 2 3 4]
Уникальные значения столбца family_status:
['женат / замужем' 'гражданский брак' 'вдовец / вдова' 'в разводе'
 'не женат / не замужем']
Уникальные значения столбца family_status_id:
[0 1 2 3 4]
Уникальные значения столбца gender:
['F' 'M' 'XNA']
Уникальные значения столбца debt:
[0 1]
Уникальные значения столбца purpose:
['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'опера

**Вывод**

Со значениями в рассматриваемых колонках все в порядке.

Ознакомимся с данными уже после замены ошибочных значений. Вызовем функцию describe() и оценим данные.

In [1048]:
data_bank_clients.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.479721    4640.441951     43.495238      0.817236   
std        0.755528    5354.870687     12.230322      0.548138   
min        0.000000      24.141633     19.000000      0.000000   
25%        0.000000     927.009265     34.000000      1.000000   
50%        0.000000    2194.220567     43.000000      1.000000   
75%        1.000000    5534.630567     53.000000      1.000000   
max        5.000000   18388.949901     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

**Вывод**
****
- Исправлены столбцы с категориальными данными - все строки приведены к одному регистру.
- Трудовой стаж с отрицательным значением переведен на положительное, слишком долгий стаж приведет к реальному. Остались вопросы по поводу возраста клиента и его стажа по категории пенсионеров. На результате исследования столбец со стажем клиента не отразится, поэтому до уточнения этих данных оставим их такими какие они есть.
- Приведено в соответствие количество детей - отрицательное заменено на положительное, а 20 к 2.
- Отсутствующий возраст заменен на средний по группам.
- Третье значение в столбце пол ('XNA') оставляем как есть, на результате не отразится.

### Заполнение пропусков медианными значениями по категории

Остались пропуски в данных о трудовом стаже и ежемесячном заработке.  

Пропущенные значения - количественные переменные. Исключим вариант удаления целых строк с пропусками, так как в этих строках все таки имеется ценная для исследования информация. Возможно заполнить пропуски нулями или средними / медианными значениям.

***Обрабатываемые данные - это статистика по уже выданным кредитам, при этом, вероятность того, что банк выдал кредиты и не поинтересовался стажем и доходом заемщика мала.***  

Чтобы исключить очень низкие или очень высокие значения стажа и дохода запоним пропуски медианными значениями по категориям клиентов.

Вполне вероятно, что значения стажа и дохода клиента по большей части зависят от типа занятости и уровня образования.

In [1049]:
data_null_employed_income = pd.DataFrame(data_bank_clients.query('days_employed.isna()'))
data_null_employed_income.head(10)

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

Рассмотрим подробднее в каких категориях по типу занятости и уровню образования образовались пропуски.  
Выведем на экран сводную таблицу. Сгруппируем пропущенные данные по типу занятости и уровню образования.

In [1050]:
pivot_null_employed_income = data_null_employed_income.pivot_table(index=['income_type', 'education'],
                                                          values=['days_employed', 'total_income'], 
                                                          aggfunc=['sum']).reset_index()
pivot_null_employed_income

income_type            education           sum             
                                         days_employed total_income
0       госслужащий               высшее           0.0          0.0
1       госслужащий            начальное           0.0          0.0
2       госслужащий  неоконченное высшее           0.0          0.0
3       госслужащий              среднее           0.0          0.0
4         компаньон               высшее           0.0          0.0
5         компаньон            начальное           0.0          0.0
6         компаньон  неоконченное высшее           0.0          0.0
7         компаньон              среднее           0.0          0.0
8         пенсионер               высшее           0.0          0.0
9         пенсионер            начальное           0.0          0.0
10        пенсионер  неоконченное высшее           0.0          0.0
11        пенсионер              среднее           0.0          0.0
12  предприниматель               высшее           0.0          0.0
13        сотрудник               высшее           0.0          0.0
14        сотрудник            начальное           0.0          0.0
15        сотрудник  неоконченное высшее           0.0          0.0
16        сотрудник              среднее           0.0          0.0

Необходимо сделать замену по 17 признакам.

Создадим сводную таблицу со средними значениями трудового стажа 'days_employed' и ежемесячного дохода 'total_income' по признаку типа занятости и уровня образования.  
Для расчета возьмем медианное значение дохода и стажа, чтобы исключить очень низкие или очень высокие доходы в единичных случаях.

In [1051]:
pivot_mean_employed_income = data_bank_clients.pivot_table(index=['income_type', 'education'],
                                                          values=['days_employed', 'total_income'], 
                                                          aggfunc=['median']).reset_index()
pivot_mean_employed_income

income_type            education        median               
                                         days_employed   total_income
0       безработный               высшее   3953.028387  202722.511368
1       безработный              среднее   3375.244668   59956.991984
2         в декрете              среднее   3296.759962   53829.130729
3       госслужащий               высшее   2531.034209  172511.107016
4       госслужащий            начальное   2787.767403  148339.290825
5       госслужащий  неоконченное высшее   1885.183639  160592.345303
6       госслужащий              среднее   2857.770974  136652.970357
7       госслужащий       ученая степень   5968.075884  111392.231107
8         компаньон               высшее   1454.659104  201785.400018
9         компаньон            начальное   1151.634460  136798.905143
10        компаньон  неоконченное высшее   1017.888120  179867.152890
11        компаньон              среднее   1670.809068  159070.690289
12        пенсионер               высшее  15256.605268  144240.768611
13        пенсионер            начальное  15011.041056  102598.653164
14        пенсионер  неоконченное высшее  15510.437569  120136.896353
15        пенсионер              среднее  15209.389119  114842.854099
16        пенсионер       ученая степень  14872.104898  177088.845999
17  предприниматель               высшее    520.848083  499163.144947
18        сотрудник               высшее   1556.457266  165640.744634
19        сотрудник            начальное   1197.176853  125994.910603
20        сотрудник  неоконченное высшее   1197.676879  151308.937846
21        сотрудник              среднее   1613.581420  136555.108821
22        сотрудник       ученая степень   2351.431934  198570.757322
23          студент               высшее    578.751554   98201.625314

Соединение двух сводных таблиц для получения средних значений трудового стажа и ежемесячного дохода только по тем категориям клиентов, в стоках которых есть пропуски в данных

In [1052]:
pivot_end = pivot_mean_employed_income.merge(pivot_null_employed_income)
pivot_end

income_type            education        median                 \
                                         days_employed   total_income   
0       госслужащий               высшее   2531.034209  172511.107016   
1       госслужащий            начальное   2787.767403  148339.290825   
2       госслужащий  неоконченное высшее   1885.183639  160592.345303   
3       госслужащий              среднее   2857.770974  136652.970357   
4         компаньон               высшее   1454.659104  201785.400018   
5         компаньон            начальное   1151.634460  136798.905143   
6         компаньон  неоконченное высшее   1017.888120  179867.152890   
7         компаньон              среднее   1670.809068  159070.690289   
8         пенсионер               высшее  15256.605268  144240.768611   
9         пенсионер            начальное  15011.041056  102598.653164   
10        пенсионер  неоконченное высшее  15510.437569  120136.896353   
11        пенсионер              среднее  15209.389119  114842.854099   
12  предприниматель               высшее    520.848083  499163.144947   
13        сотрудник               высшее   1556.457266  165640.744634   
14        сотрудник            начальное   1197.176853  125994.910603   
15        сотрудник  неоконченное высшее   1197.676879  151308.937846   
16        сотрудник              среднее   1613.581420  136555.108821   

             sum               
   days_employed total_income  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  
5            0.0          0.0  
6            0.0          0.0  
7            0.0          0.0  
8            0.0          0.0  
9            0.0          0.0  
10           0.0          0.0  
11           0.0          0.0  
12           0.0          0.0  
13           0.0          0.0  
14           0.0          0.0  
15           0.0          0.0  
16           0.0          0.0

In [1053]:
# создадим 4 списка, значения каждого будут соответствовать значениям в сводной таблице pivot_end
# выведем на экран полученные списки и длину одного из них (для проверки)

median_it, median_ed, median_de, median_ti = [], [], [], []

for i in range(len(pivot_end)):
    median_it.append(pivot_end['income_type'][i]) # список с типом занятости
    median_ed.append(pivot_end['education'][i])   # список с образованием
    median_de.append(pivot_end['median']['days_employed'][i]) # медианное значение стажа по категории
    median_ti.append(pivot_end['median']['total_income'][i])  # медианное значение дохода по категории
print(median_it, median_ed, median_de, median_ti, sep='\n')
print(len(median_it))

['госслужащий', 'госслужащий', 'госслужащий', 'госслужащий', 'компаньон', 'компаньон', 'компаньон', 'компаньон', 'пенсионер', 'пенсионер', 'пенсионер', 'пенсионер', 'предприниматель', 'сотрудник', 'сотрудник', 'сотрудник', 'сотрудник']
['высшее', 'начальное', 'неоконченное высшее', 'среднее', 'высшее', 'начальное', 'неоконченное высшее', 'среднее', 'высшее', 'начальное', 'неоконченное высшее', 'среднее', 'высшее', 'высшее', 'начальное', 'неоконченное высшее', 'среднее']
[2531.0342091733646, 2787.7674025586484, 1885.1836391786355, 2857.7709737955824, 1454.6591042048294, 1151.634459637238, 1017.888120007462, 1670.809067562865, 15256.605267845207, 15011.041056265909, 15510.437569187496, 15209.389119446227, 520.8480834953765, 1556.4572664208927, 1197.1768530446493, 1197.6768793542578, 1613.581419770452]
[172511.10701550177, 148339.29082488286, 160592.34530334338, 136652.9703567915, 201785.40001791567, 136798.9051432623, 179867.15288993614, 159070.69028882936, 144240.76861073746, 102598.653

In [1054]:
# проверим правильность заполнения списков.
print(median_it[0], median_ed[0], median_de[0], median_ti[0])
print(median_it[4], median_ed[4], median_de[4], median_ti[4])
print(median_it[10], median_ed[10], median_de[10], median_ti[10])
print(median_it[15], median_ed[15], median_de[15], median_ti[15])

госслужащий высшее 2531.0342091733646 172511.10701550177
компаньон высшее 1454.6591042048294 201785.40001791567
пенсионер неоконченное высшее 15510.437569187496 120136.89635252525
сотрудник неоконченное высшее 1197.6768793542578 151308.9378458309


Значения в списках соответствую значениям в сводной таблице pivot_end.  
Индекс значения в списке соответствует индексу строки в pivot_end.  


Можно приступать к замене пропущенных значений в колонках 'days_employed' и 'total_income'

In [1055]:
for i in range(len(median_it)):
    data_bank_clients.loc[(data_bank_clients['income_type'] == median_it[i]) & 
                          (data_bank_clients['education'] == median_ed[i]) &
                          (data_bank_clients['days_employed'].isna()),
                          'days_employed'] = median_de[i]

In [1056]:
for i in range(len(median_it)):
    data_bank_clients.loc[(data_bank_clients['income_type'] == median_it[i]) & 
                          (data_bank_clients['education'] == median_ed[i]) &
                          (data_bank_clients['total_income'].isna()),
                          'total_income'] = median_ti[i]

Посчитаем количество пропусков после замены пропущенных значений

In [1057]:
data_bank_clients.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Все пропуски были заменены на медианные значения трудового стажа и дохода в соответствии с типом занятости и уровнем образования заемщика

In [1058]:
data_bank_clients.pivot_table(index=['income_type', 'education'],
                              values=['days_employed', 'total_income'], 
                              aggfunc=['median']).reset_index()

income_type            education        median               
                                         days_employed   total_income
0       безработный               высшее   3953.028387  202722.511368
1       безработный              среднее   3375.244668   59956.991984
2         в декрете              среднее   3296.759962   53829.130729
3       госслужащий               высшее   2531.034209  172511.107016
4       госслужащий            начальное   2787.767403  148339.290825
5       госслужащий  неоконченное высшее   1885.183639  160592.345303
6       госслужащий              среднее   2857.770974  136652.970357
7       госслужащий       ученая степень   5968.075884  111392.231107
8         компаньон               высшее   1454.659104  201785.400018
9         компаньон            начальное   1151.634460  136798.905143
10        компаньон  неоконченное высшее   1017.888120  179867.152890
11        компаньон              среднее   1670.809068  159070.690289
12        пенсионер               высшее  15256.605268  144240.768611
13        пенсионер            начальное  15011.041056  102598.653164
14        пенсионер  неоконченное высшее  15510.437569  120136.896353
15        пенсионер              среднее  15209.389119  114842.854099
16        пенсионер       ученая степень  14872.104898  177088.845999
17  предприниматель               высшее    520.848083  499163.144947
18        сотрудник               высшее   1556.457266  165640.744634
19        сотрудник            начальное   1197.176853  125994.910603
20        сотрудник  неоконченное высшее   1197.676879  151308.937846
21        сотрудник              среднее   1613.581420  136555.108821
22        сотрудник       ученая степень   2351.431934  198570.757322
23          студент               высшее    578.751554   98201.625314

**Вывод**
***

Все пропущенные значения были обработаны и заполнены медианными значениями стажа и дохода в зависимости от категории занятости и уровня образования заемщика.

### Замена типа данных

Для удобства дальнейшей работы с обработаем типы данных столбцов.

In [1059]:
data_bank_clients.dtypes

children              int64
days_employed       float64
dob_years             int64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
income_type          object
debt                  int64
total_income        float64
purpose              object
dtype: object

Все типы данных  в столбцах соответствуют представленной в них информации. Где числа - там тип int64 или float64, где значения строчные - object. Категоральные значения, выраженные числами (столбцы 'education_id', 'family_status_id', 'debt') представлены целочисленным типом данных.

В столбцах days_employed (трудовой стаж) и total_income (ежемесячный доход) тип данных float64 и много знаков после запятой, что усложняет читаемость данных и дальнейшую работу с ними.  

Для того, чтобы привести значения с плавающей точкой к целочисленным можно применить метод astype(). Так как приводить к целочисленному типу int нужно две колонки воспользуемся словарем.

Приведем столбцы 'days_employed' и 'total_income' к целочисленному типу и выведем измененные типы на экран методом dtypes()

In [1060]:
data_bank_clients = data_bank_clients.astype({'days_employed': 'int64', 'total_income' : 'int64'})
data_bank_clients.dtypes

children             int64
days_employed        int64
dob_years            int64
education           object
education_id         int64
family_status       object
family_status_id     int64
gender              object
income_type         object
debt                 int64
total_income         int64
purpose             object
dtype: object

**Вывод** 
***
Типы данных в столбцах 'days_employed' (трудовой стаж) и 'total_income' (ежемесячный доход) изменены с типов значений с павающей точкой (float64) на целочисленные (int64) для удобства дальнейшей работы с ними.

Других проблем с типами данных в датафрейме нет.

### Обработка дубликатов

Найдем количество дубликатов в датафрейме, если такие имеются.
Вызовем метод duplicated() и sum() для подсчета дублированных строк.

In [1061]:
print('Дубликатов в датафрейме: ', data_bank_clients.duplicated().sum())

Дубликатов в датафрейме:  71


Уникального идентификатора заемщика или какого - либо другого признака по которому возможно было бы выделить уникальных клиентов и уже на них проверить дубликаты в таблице нет, поэтому подсчитать количество дубликатов по определенному признаку не представляется возможным: по любым признакам имеющимся в нашем распоряжении дубликаты будут, например:

In [1062]:
print(data_bank_clients['income_type'].duplicated().value_counts())

True     21517
False        8
Name: income_type, dtype: int64


Теоретически, дубликаты могли появится в результате технического сбоя при при выгрузке данных. Так как уникального идентификатора нет, уточнить относятся ли наденные дубликаты к однуму клиенту не представляется возможным. 

Для дальнейшей обработки удалим дубликаты и заменим индексы и проверим количество дубликатов после удаления

In [1063]:
data_bank_clients = data_bank_clients.drop_duplicates().reset_index(drop=True)
print('Дубликатов в датафрейме: ', data_bank_clients.duplicated().sum())

Дубликатов в датафрейме:  0


**Вывод**
***
- всего был найден 71 дубликат в датафрейме методом duplicated().sum()  
- методом drop_duplicates() дубликаты в строках были удалены

### Лемматизация

Найдем наиболее часто встречающиеся слова в категории целей кредита. В остальных столбцах уже можно выделить отдельные категории, их лемматизировать не нужно.  
Воспользуемся библиотекой pymystem3 с функцией лемматизации Mystem()

In [1064]:
m = Mystem()

Подготовим столбец с целями кредита для лемматизации и выявления наиболее часто встречающихся слов.

Рассмотрим подробнее уникальные значения столбца 'purpoze' - целей кредита и количество их повторений для этого воспользуемся методом value_counts(). Сохраним в purpose_unique_dbc уникальные значения столбца и переведем их в строку.

In [1065]:
display(data_bank_clients['purpose'].value_counts())
purpose_unique_dbc = ','.join(data_bank_clients['purpose'].unique())

свадьба                                   791
на проведение свадьбы                     768
сыграть свадьбу                           765
операции с недвижимостью                  675
покупка коммерческой недвижимости         661
операции с жильем                         652
покупка жилья для сдачи                   651
операции с коммерческой недвижимостью     650
покупка жилья                             646
жилье                                     646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          624
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

Лемматизируем слова в переменной purpose_unique_dbc и подсчитаем количество слов и отсортируем по убыванию

In [1066]:
lemme_purpose = m.lemmatize((purpose_unique_dbc))
count_lemme_purpose = Counter(lemme_purpose)
count_lemme_purpose.most_common()

[(' ', 59),
 (',', 37),
 ('покупка', 10),
 ('недвижимость', 10),
 ('автомобиль', 9),
 ('образование', 9),
 ('жилье', 7),
 ('с', 5),
 ('операция', 4),
 ('на', 4),
 ('свой', 4),
 ('свадьба', 3),
 ('строительство', 3),
 ('получение', 3),
 ('высокий', 3),
 ('дополнительный', 2),
 ('для', 2),
 ('коммерческий', 2),
 ('жилой', 2),
 ('подержать', 2),
 ('заниматься', 2),
 ('сделка', 2),
 ('приобретение', 1),
 ('сыграть', 1),
 ('проведение', 1),
 ('семья', 1),
 ('собственный', 1),
 ('со', 1),
 ('профильный', 1),
 ('сдача', 1),
 ('ремонт', 1),
 ('\n', 1)]

In [1067]:
sorted(list(count_lemme_purpose))

['\n',
 ' ',
 ',',
 'автомобиль',
 'высокий',
 'для',
 'дополнительный',
 'жилой',
 'жилье',
 'заниматься',
 'коммерческий',
 'на',
 'недвижимость',
 'образование',
 'операция',
 'подержать',
 'покупка',
 'получение',
 'приобретение',
 'проведение',
 'профильный',
 'ремонт',
 'с',
 'свадьба',
 'свой',
 'сдача',
 'сделка',
 'семья',
 'со',
 'собственный',
 'строительство',
 'сыграть']

Проведем подсчет лемм по всем значениям в столбце целей - purpose и сохраним в purpose_dbc_all все значения столбца purpose и переведем их в строку

In [1068]:
purpose_dbc_all = []
for i in range(len(data_bank_clients['purpose'])):
    purpose_dbc_all.append(data_bank_clients['purpose'][i])
purpose_dbc_all = ','.join(purpose_dbc_all)

Лемматизируем слова в столбце purpose и подсчитаем количество слов во всем столбце и отсортируем по убыванию

In [1069]:
lemme_purpose_all = m.lemmatize(purpose_dbc_all)
count_lemme_purpose_all = Counter(lemme_purpose_all)
count_lemme_purpose_all.most_common()

[(' ', 33570),
 (',', 21453),
 ('недвижимость', 6351),
 ('покупка', 5897),
 ('жилье', 4460),
 ('автомобиль', 4306),
 ('образование', 4013),
 ('с', 2918),
 ('операция', 2604),
 ('свадьба', 2324),
 ('свой', 2230),
 ('на', 2222),
 ('строительство', 1878),
 ('высокий', 1374),
 ('получение', 1314),
 ('коммерческий', 1311),
 ('для', 1289),
 ('жилой', 1230),
 ('сделка', 941),
 ('дополнительный', 906),
 ('заниматься', 904),
 ('подержать', 853),
 ('проведение', 768),
 ('сыграть', 765),
 ('сдача', 651),
 ('семья', 638),
 ('собственный', 635),
 ('со', 627),
 ('ремонт', 607),
 ('приобретение', 461),
 ('профильный', 436),
 ('подержанный', 111),
 ('\n', 1)]

Объединим все цели в 4 категории, сохраним в список

1. Все сделки с недвижимостью, ремонт, сдача жилья, строительство объединим в одну категорию - недвижимость.
2. Все, что связано с автомобилями - автомобиль.
3. Дополнительное образование, высшее образование - образование.
4. Сыграть свадьбу - свадьба.

Выделим общий список категорий целей кредита - category_purpose.

In [1070]:
category_purpose = ['автомобиль', 'недвижимость', 'образование', 'свадьба']

**Вывод**
***
Путем лемматизации столбца 'purpose' (целей выдачи кредита) было выделено 4 основные целевые категории: 'автомобиль', 'недвижимость', 'образование', 'свадьба'

### Категоризация данных

Категоризируем данные по цели выдачи кредитов. Для получения ответов на вопрос как влияет цель кредита на его возврат в срок.
В предыдущем разделе было выделено 4 больших категории целей для выдачи кредитов:
- 'автомобиль', 
- 'недвижимость',
- 'образование',
- 'свадьба'

Добавим новый столбец с категорией цели кредита

In [1071]:
def f_purpose_type(purpose):
    if 'автомобил' in purpose:
        pt = 'автомобиль'
    elif 'образован' in purpose:
        pt = 'образование'
    elif 'свад' in purpose:
        pt = 'свадьба'
    elif 'жил' in purpose or 'недвиж' in purpose:
        pt = 'недвижимость'
    else:
        pt = 'пропущена'
    return pt      

In [1072]:
# заполним значения столбца 'purpose_type' с помощью функции f_purpose_type
data_bank_clients['purpose_type'] = data_bank_clients['purpose'].apply(f_purpose_type)

In [1073]:
# проверим правильность заполнения - выведем уникальные значения столбца 'purpose_type'
data_bank_clients['purpose_type'].unique()

array(['недвижимость', 'автомобиль', 'образование', 'свадьба'],
      dtype=object)

Проверка на ошибку работы функции прошла успешно. Категория 'пропущена' отсутствует.


In [1074]:
# сгруппируем данные по большой цели и по исходной цели
data_bank_clients_purpose = data_bank_clients.pivot_table(index=['purpose_type', 'purpose'],
                              values=['debt'], 
                              aggfunc=['count']).reset_index()
data_bank_clients_purpose.head()

purpose_type                             purpose count
                                                    debt
0   автомобиль                          автомобили   478
1   автомобиль                          автомобиль   494
2   автомобиль               на покупку автомобиля   471
3   автомобиль  на покупку подержанного автомобиля   478
4   автомобиль        на покупку своего автомобиля   505

***Категоризируем данные по уровню дохода***

Категоризация доходов понадобится для ответа на вопрос как влияет уровень дохода на возврат кредитов в срок.

In [1075]:
# определим какой вообще имеется доход у клиентов
total_income_category = data_bank_clients['total_income'].describe().astype('int')
total_income_category

count      21454
mean      165453
std        98288
min        20667
25%       107515
50%       143707
75%       198307
max      2265604
Name: total_income, dtype: int64

Разделим на 4 группы:
1. Низкий доход - менее 25 %
2. Средний доход - от 25 % до 50 %
3. Высокий доход от 50 % до 75 %
4. Очень высокий доход - выше 75 %

In [1080]:
# разделим доход клиентов на 4 основные категории:
data_bank_clients['category_income_type'] = pd.qcut(data_bank_clients['total_income'], 4, 
                               labels=["низкий доход", "средний доход", "высокий доход", "очень высокий доход"])

In [1081]:
# проверим правильность заполнения - выведем уникальные значения столбца 'category_income_type'
data_bank_clients['category_income_type'].unique()

[очень высокий доход, средний доход, высокий доход, низкий доход]
Categories (4, object): [низкий доход < средний доход < высокий доход < очень высокий доход]

Пропущенных категорий по доходам не появилось. Все данные столбца 'total_income' категоризированы для последующей обработки.

**Вывод**
***
- Путем категоризации столбца целей кредита 'purpose' был добавлен новый столбец 'purpose_category', который объединил разные варианты написания целей кредита по основным категориям: недвижимость, образование, свадьба и автомобиль.
- Данные по доходам были объединены по категориям: 'очень высокий доход', 'средний доход', 'высокий доход', 'низкий доход' и добавлен новый столбец 'category_income_type'

**ВЫВОДЫ ПО ГЛАВЕ 2. Предобработка данных**
***
Найденные артефакты обработаны.  

Пропущенные значения в колонках с трудовым стажем и доходом были заменены на медианные значения в соответствии с типом занятости и уровнем образования.

Необходимо уточнить по какому принципу собиралась информация о трудовом стаже заемщиков, так как после перевода значений, остались строки, в которых стаж клиента слишком высок для его возраста.  
В данном исследовании трудовой стаж не повлияет на результат поэтому, до уточнения данных оставим их такими какие они есть.

Лемматизирован столбец с целями кредита. 

Категоризированы данные по целям кредита и уровню дохода. Результат два новых столбца - 'purpose_type' и 'category_income_type'.

##  Результаты исследования

Результатом исследования будут ответы на следующие вопросы:
1. Есть ли зависимость между наличием детей и возвратом кредита в срок?
2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
4. Как разные цели кредита влияют на его возврат в срок?

Для ответов на эти вопросы подготовим функцию, которая будет группировать и подсчитывать данные по интересующим нас столбцам.

share_percent(column): 
- функция нахождения процента данных по категории от общих данных
- параметр функции - наименование столбца по которому необходимо получить данные
- выход - % данных по категориям столбца

In [1082]:
def share_percent(column):
    count_dbc = (data_bank_clients[column].value_counts()).to_frame()
    count_dbc['share_percent'] = round((data_bank_clients[column].value_counts() / len(data_bank_clients) * 100), 2)
    return count_dbc

share_debt(column): 
- функция для подсчета процента задолженностей
- параметр функции - наименование столбца по которому необходимо получить данные
- выход - сводная таблица с колонками: '0' - количество клиентов без задолженности,
- '1' - количество клиентов с задолженностью, 'total' - общее количество клиентов по группам
- 'share_bad_debt_%' - процент клиентов с задолженностью по категориям

In [1]:
def share_debt(column):
    data_bank_clients_debt = data_bank_clients.pivot_table(index=[column], 
                                                                 columns=['debt'], values='purpose', aggfunc='count').fillna(0)
    data_bank_clients_debt['total'] = data_bank_clients_debt[0] + data_bank_clients_debt[1]
    data_bank_clients_debt['share_bad_debt_%'] = round((data_bank_clients_debt[1] / 
                                                        data_bank_clients_debt['total'] * 100), 2)
    return data_bank_clients_debt

### Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [1084]:
share_percent('children')

children  share_percent
0     14091          65.68
1      4855          22.63
2      2128           9.92
3       330           1.54
4        41           0.19
5         9           0.04

Выборка для заемщиков с количеством детей от 3 до 5 слишком мала для оценки зависимости между наличием детей и возвратом кредита в срок  
(совокупно < 2 % от общего числа заемщиков).

In [1085]:
share_debt('children').sort_values(by='share_bad_debt_%')

debt            0       1    total  share_bad_debt_%
children                                            
5             9.0     0.0      9.0              0.00
0         13028.0  1063.0  14091.0              7.54
3           303.0    27.0    330.0              8.18
1          4410.0   445.0   4855.0              9.17
2          1926.0   202.0   2128.0              9.49
4            37.0     4.0     41.0              9.76

**Вывод**
***

- Бездетные клиенты примерно на 2% лучше отдают кредиты по сравнению с клентами с одним или двумя детьми. Процент наличия задолженностей у них ниже.

- Практически на одном уровне по количеству задолженностей находятся клиенты с одним или двумя детьми (с разницей ~0.3%).  

- Выборка по клиентам с 4 и 5 детьми слишком мала для оценки. 

- Стоит заметить, что самый высокий процент задолженностей имеют клиенты с одним, двумя и четырьмя детьми.



### Есть ли зависимость между семейным положением и возвратом кредита в срок?

Рассмотрим подробнее категорию семейного положения и оценим долю каждой категории

In [1086]:
share_percent('family_status')

family_status  share_percent
женат / замужем                12339          57.51
гражданский брак                4151          19.35
не женат / не замужем           2810          13.10
в разводе                       1195           5.57
вдовец / вдова                   959           4.47

Создадим сводную таблицу в которой строки - семейное положение клиентов  
'share_bad_debt' - процент просроченных кредитов в соответствии с категорией

In [1087]:
share_debt('family_status').sort_values(by='share_bad_debt_%')

debt                       0    1  total  share_bad_debt_%
family_status                                             
вдовец / вдова           896   63    959              6.57
в разводе               1110   85   1195              7.11
женат / замужем        11408  931  12339              7.55
гражданский брак        3763  388   4151              9.35
не женат / не замужем   2536  274   2810              9.75

**Вывод**
***

- Процент задолженностей у людей в браке и в разводе примерно одинаковый. Кредит возвращают ~ 93 % клиентов.  


- Самый высокий процент просроченных кредитов у людей не состоявших в браке (гражданский брак или не женат/не замужем). По этим категориям просроченные задолженности отсутствуют у ~ 90 % клиентов.  


- Среди вдовцов самый низкий процент просрочек. При этом в данных эта категория представлена в небольшом (относительно других) количестве.  

Резкой зависимости между семейным положением и возвратом кредита в срок не наблюдается: процент задолженностей в категориях варьируется от ~7% до ~10%. Нельзя сказать, что, например, задолженность по кредитам имелась только у людей в разводе.  
При этом некоторая зависимость прослеживается: с разницей до ~3% клиенты в браке/разводе/вдовцы имеют меньше задолженностей, чем клиенты в гражданском браке/не в браке.

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Рассмотрим подробнее категории доходов заемщиков и оценим долю каждой категории:
- в строках - категория доходов, 
- в стобце 'category_income_type' - количество клиентов с соответствующим доходом
- в столбце 'share_percent' - процент категории клиентов от общего числа

In [1088]:
share_percent('category_income_type')

category_income_type  share_percent
очень высокий доход                  5364           25.0
низкий доход                         5364           25.0
высокий доход                        5363           25.0
средний доход                        5363           25.0

Создадим сводную таблицу в которой строки - уровень дохода
- 'share_bad_debt' - процент просроченных кредитов в соответствии с категорией

In [1089]:
share_debt('category_income_type').sort_values(by='share_bad_debt_%')

debt                     0    1  total  share_bad_debt_%
category_income_type                                    
очень высокий доход   4991  373   5364              6.95
низкий доход          4937  427   5364              7.96
высокий доход         4907  456   5363              8.50
средний доход         4878  485   5363              9.04

**Вывод**
***

- Самый высокий процент просроченных выплат по кредиту у клиентов со средним  и высоким доходом. 
- Самое низкое количество просрочек имеют люди с очень высоким доходом.
- Клиенты с низкими доходами менее склонны к просрочкам, по сравнению с клиентами со средним и высоким доходами.

От уровня дохода зависит возврат кредита в срок. Чаще задолженности происходят у людей со средним и высоким уровнем дохода. Люди с очень высоким доходом имеют меньше всего просрочек.

### Как разные цели кредита влияют на его возврат в срок?

Рассмотрим подробнее категории целей и оценим долю каждой:
- в строках - цель кредита
- в стобце 'purpose_type' - количество клиентов с целью кредита
- в столбце 'share_percent' - процент категории клиентов от общего числа

In [1090]:
share_percent('purpose_type')

purpose_type  share_percent
недвижимость         10811          50.39
автомобиль            4306          20.07
образование           4013          18.71
свадьба               2324          10.83

Создадим сводную таблицу в которой строки - цель кредита
- 'share_bad_debt' - процент просроченных кредитов в соответствии с категорией

In [1091]:
share_debt('purpose_type').sort_values(by='share_bad_debt_%')

debt              0    1  total  share_bad_debt_%
purpose_type                                     
недвижимость  10029  782  10811              7.23
свадьба        2138  186   2324              8.00
образование    3643  370   4013              9.22
автомобиль     3903  403   4306              9.36

**Вывод**
***
- Цели кредита влияют на количество задолженностей по кредиту.
Так, люди, которые берут кредит на различные сделки или манипуляции с недвижимостью допускают на ~2% меньше просрочек, чем те, кто берет кредит на образование или на покупку автомобиля.  
- Из 4 категорий целей кредита - автомобиль показывает самый высокий результат по задолженностям. Кредит на образования идет следом с разницей ~ 0.15%. Порядка 9% заемщиков данным целевым категорям имеют задолженности.

## Общие выводы

Подводя итоги нашего исследования, отметим что прослеживается некоторая зависимость возврата кредита в срок от наличия и количества детей, семейного положения, дохода и целей кредита:

- У бездетных клиентов задолженностей перед банком меньше, чем у клиентов с 1 или 2 детьми.
- При этом у людей не состоящих в официальном браке процент задолженностей выше, чем у людей в браке или разводе.
- Меньше задолженностей по кредиту имеют люди с очень высокими или низкими доходам по сравнению с клиентами с высокими или средними доходами.
- Самый рисковая цель получения кредита - это автомобиль. Рядом с ней и образование. Меньше всего задолженностей у людей, которые берут кредит на сделки с недвижимостью.

Итак, наименее рисковый заемщик для банка - бездетный клиент в браке или разводе с очень высоким доходом с целью приобрести недвижимость.
А наиболее рисковый - клиент с 1-2 детьми в неофициальном браке со средним доходом и целью приобретения автомомбиля.


## Чек-лист готовности проекта

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.